text/000-buildpacks-contribute-default-process-type.md

```diff
@@ -0,0 +1,133 @@
+
+process type - a CNB construct representing a process that can be started by the launcher. A process type has the following properties as described [here](https://github.com/buildpacks/spec/blob/main/buildpack.md#launchtoml-toml): type (string), command (string), arguments (array of string), direct (boolean). Buildpacks declare process types during the `build` phase by writing entries into `<layers>/launch.toml`.
+
+default process type - describes the process type that will be started by the launcher when the app image is run. It can be overriden by overriding the image entrypoint. Example: an image with default process type of web would have an entrypoint of `/cnb/process/web` which is a symlink that points to `/cnb/lifecycle/launcher`. The launcher uses argv0 to determine which process to start.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this? Easier and more intuitive for end users.
+- What use cases does it support? Apps with one or more process type(s).
+- What is the expected outcome? The experience should be seamless for end users. Buildpacks should produce a sensible default process type without the user having to know or care that process types exist. The platform should still be able to override the default process type by passing `-process-type` to the lifecycle.
+
+# What it is
+[what-it-is]: #what-it-is
+
+#### Happy path
+
+Example: single process app with a buildpack declaring a `default-process` of type `web` -> the app image will have a default process type of `web`
+
+Example: single process app with no buildpacks declaring a `default-process` -> the app image will have no default process. Users must specify the process to run at runtime. (This assumes https://github.com/buildpacks/spec/pull/137)
+
+Example: single process app with no buildpacks declaring a `default-process` and the user passes `-process-type worker` -> the app image will have a default process type of `worker`
+
+Example: multi-process app with a buildpack declaring a `default-process` of type `web` -> the app image will have a default process type of `web`
+
+#### Overrides
+
+Example: multi-process app with an earlier buildpack declaring a `default-process` of type `web` and a later buildpack declaring a `default-process` of type `worker` -> the app image will have a default process type of `worker`
+
+Example: multi-process app with an earlier buildpack declaring a `default-process` of type `web` and a later buildpack declaring a `default-process` of type `worker` and `override = false` -> the app image will have a default process type of `web`
+
+Example: multi-process app with a buildpack declaring a `default-process` of type `web` and the user passes `-process-type worker` -> the app image will have a default process type of `worker`
+
+#### Overrides - edge cases
```

Comment:
Should we warn when these occur?

@sclevine added this section and https://github.com/buildpacks/rfcs/pull/110/files#diff-419c1902fa5125d1c8e9a11131832217R97 to address some of your concerns from WG.

Reply:
> Should we warn when these occur?

Just my two cents..

> Example: multi-process app with a buildpack declaring a process type worker and a default-process of type web -> the app image will have no default process. Users must specify the process to run at runtime.

I think this one ^ should be an error case, assuming the buildpack attempting to set the default did not declare a `web` process type at all.

> Example: multi-process app with an earlier buildpack declaring a default-process of type web and a later buildpack redefining what web means -> the app image will have no default process. Users must specify the process to run at runtime.

I think this one ^ should be a warning.

> Example: multi-process app with an earlier buildpack declaring a default-process of type web and a later buildpack redefining what web means while also declaring web as the default-process -> the app image will have a default process type of web

And this last one should succeed without warning.

Reply:
Do we want to add this to the RFC or leave this out?

Reply:
Now that `default` is a property of the process, the first case (error case) isn't relevant anymore.

Noted in the RFC that the lifecycle should print a warning for the second case:

> Example: multi-process app with an earlier buildpack declaring a default-process of type web and a later buildpack redefining what web means -> the app image will have no default process. Users must specify the process to run at runtime.